WIP : Elektron Logging library

balandi1 opened the pull request.

balandi1 committed:
Initial elektromLogging library

balandi1 committed:
Changes made everywhere inorder to use elektron logging library for logging

balandi1 committed:
Changes made everywhere inorder to use elektronLogging library for logging

balandi1 committed:
Replaced old usage with new library

balandi1 committed:
Deleted the unused logging library

balandi1 committed:
Fixed minor bug

pradykaushik commented:
@ridv this PR corresponds to retrofitting code to use logrus for logging and get rid of the built-in logging library.

pradykaushik commented:
@balandi1 please also make a PR to [elektron-vendor](https://github.com/spdfg/elektron-vendor) with the updated dependencies to make sure that dependencies are kept up to date.

ridv commented on def/taskUtils.go:
```diff
@@ -21,11 +21,13 @@ package def
 	"github.com/montanaflynn/stats"
+     "github.com/spdfg/elektron/elektronLogging"
+	elekLogT "github.com/spdfg/elektron/elektronLogging/types"
+    log "github.com/sirupsen/logrus"
```
Format error, fix spacing or run through [goimports](https://godoc.org/golang.org/x/tools/cmd/goimports)

balandi1 replied:
Yes, I will fix all the formatting errors

ridv commented on def/taskUtils.go:
```diff
@@ -21,11 +21,13 @@ package def
 	"github.com/mash/gokmeans"
 	"github.com/montanaflynn/stats"
+     "github.com/spdfg/elektron/elektronLogging"
+	elekLogT "github.com/spdfg/elektron/elektronLogging/types"
```
Let's pick a better alias, this doesn't really say much about what this package holds.

balandi1 replied:
Okay. Will changing it to `elekLogTypes` or `logTypes` simply work ?

ridv replied:
logTypes would be fine

ridv commented on def/taskUtils.go:
```diff
@@ -105,7 +108,8 @@ func clusterSizeAvgMMMPU(tasks []Task, taskObservation func(task Task) []float64
 				// skip this value
 				// there is an error in the task config.
-				log.Println(err)
+				elektronLogging.ElektronLog.Log(elekLogT.ERROR, log.ErrorLevel, 
```
Is  Logf an option here instead of sprinting?

balandi1 replied:
Yes. I will work on implementing Logf()

ridv commented on elektronLogging/ElektronFormatter.go:
```diff
@@ -0,0 +1,55 @@
+	"bytes"
+	"fmt"
+	"strings"
+    "github.com/fatih/color"
```
Formatting

balandi1 replied:
Done

ridv commented on elektronLogging/ElektronFormatter.go:
```diff
@@ -0,0 +1,55 @@
+		b = &bytes.Buffer{}
+	}
+
+    levelColor := f.getColor(entry)
```
Formatting

balandi1 replied:
Done

ridv commented on elektronLogging/ElektronFormatter.go:
```diff
@@ -0,0 +1,55 @@
+	level := levelColor.Sprintf("[%s]:",strings.ToUpper(entry.Level.String()))
+    message := fmt.Sprintf("%s %s ",level,entry.Time.Format(f.TimestampFormat))
+    if entry.Message != "" {
+	    message = fmt.Sprintf("%s %s %s ",level,entry.Time.Format(f.TimestampFormat), entry.Message)
```
Use [strings.join](https://golang.org/pkg/strings/#Join) instead. You should create the slice and append the message if necessary.

balandi1 replied:
Okay. Sure

ridv commented on elektronLogging/ElektronFormatter.go:
```diff
@@ -0,0 +1,55 @@
+	var formattedFields []string
+	for key, value := range entry.Data {
+		formattedFields = append(formattedFields, 
+		strings.Join([]string{key, fmt.Sprintf("%s", value)}, "="))
```
This seems really inefficient. Wouldn't value already have to be a string for this to succeed? In which case, I don't think we gain anything here by suing a sprintf.

balandi1 replied:
value is not a string here, it is of type interface{}. But to avoid using sprintf, would it be fine to use type assertion? For eg : `value.(string) `

pradykaushik replied:
Peeked at how [logrus](https://github.com/sirupsen/logrus/blob/67a7fdcf741f4d5cee82cb9800994ccfd4393ad0/text_formatter.go#L315) serializes values and seems like fmt.Sprint is used if type assertion fails.

ridv replied:
ahh alright, I thought it was a string coming in. Alright, if it's been sanity checked then i'm good with it.

ridv commented on elektronLogging/clsfnTaskDistOverheadLogger.go:
```diff
@@ -0,0 +1,49 @@
+}
+
+func NewClsfnTaskDistOverheadLogger(logType int, prefix string) *ClsfnTaskDistOverheadLogger {
+	cLog := new(ClsfnTaskDistOverheadLogger)
```
Use struct literals instead of new as it's more descriptive:
`&ClsfnTaskDistOverheadLogger{}`

balandi1 replied:
Okay, will do the changes

ridv commented on elektronLogging/clsfnTaskDistOverheadLogger.go:
```diff
@@ -0,0 +1,49 @@
+func (cLog *ClsfnTaskDistOverheadLogger) SetLogFile(prefix string) {
+
+	tskDistLogPrefix := prefix + config.TaskDistConfig.FilenameExtension
+	if logDir != "" {
```
Where is log dir coming from? Shouldn't be a global

balandi1 replied:
It is defined in `createLogDir.go`. Since all these files belong to same package, I accessed it in this way.

pradykaushik replied:
I think it will be better if it is encapsulated.
Something like
```go
type logDir struct {
    name string
    // other fields if necessary.
}
```

balandi1 replied:
Done

ridv replied:
looks good now thanks!

ridv commented on elektronLogging/createLogDir.go:
```diff
@@ -0,0 +1,38 @@
+func createLogDir(prefix string, startTime time.Time) string {
+	// Creating directory to store all logs for this run
+	logDirName := "./" + prefix + strconv.Itoa(startTime.Year())
+	logDirName += "-"
```
Why not use `string.join([]string{...}, "")` ?

balandi1 replied:
Yeah, thats right

ridv commented on elektronLogging/loggerChain.go:
```diff
@@ -0,0 +1,37 @@
+	}
+}
+
+/*func cloneFields(logData data.LogData) log.Fields {
```
If the code isn't needed it shouldn't be left here. If the commented out code is still needed there should be some justification for it still being here.

balandi1 replied:
Okay. Will remove the unnecessary code

ridv reviewed (changes requested):
Run code all modified code through goimports

balandi1 committed:
Code formatting

balandi1 committed:
Used string join at neccesary places. Removed unused code

pradykaushik commented:
Just checked the code and made a test run.
1. Seems like the log prefix provided using the `-logPrefix` commandline option is not used when creating the log directory.
2. Color codes should be removed from the `ESC[32;1m[INFO]:ESC[0m ` prefix on each line in the logs.
2. PCP log file has `ESC[32;1m[INFO]:ESC[0m ` prefix on each line. PCP logs should not have this prefix and should just have the timestamp and the message. One way of making this generic is by supporting overriding of log config for each element in the chain. For example, PCP logs are to be persisted in a different format than the universal format specified for all others.
3. If a log type is disabled in (specifying `enabled: false` in the logConfig yaml file), then the corresponding log file should not be created. This would result in no log file being empty.

pradykaushik commented on def/taskUtils.go:
```diff
@@ -1,31 +1,33 @@
 	"github.com/mash/gokmeans"
 	"github.com/montanaflynn/stats"
+	log "github.com/sirupsen/logrus"
+	"github.com/spdfg/elektron/elektronLogging"
```
I would alias this to _elekLog_ as you seem to be using that prefix when aliasing imports of sub-packages.

balandi1 replied:
Okay. Done

pradykaushik commented on def/taskUtils.go:
```diff
@@ -50,7 +52,8 @@ func (tc TasksToClassify) taskObservationCalculator(task Task) []float64 {
 		return []float64{task.Watts}
 	} else {
-		log.Fatal("Unable to classify tasks. Missing Watts or ClassToWatts attribute in workload.")
+		elektronLogging.ElektronLog.Log(elekLogT.ERROR, log.FatalLevel,
```
This is very confusing. In fact, now that we are using logrus, we can get rid of the following types in `types.go`.
1. ERROR
2. GENERAL
3. SUCCESS
4. WARNING

Instead, we should just mention that we are logging to the console. The log level is anyway specified as the second argument.

balandi1 replied:
Yeah, that's right.

pradykaushik commented on def/taskUtils.go:
```diff
@@ -105,7 +108,8 @@ func clusterSizeAvgMMMPU(tasks []Task, taskObservation func(task Task) []float64
 				// skip this value
 				// there is an error in the task config.
-				log.Println(err)
+				elektronLogging.ElektronLog.Log(elekLogT.ERROR, log.ErrorLevel,
```
Replace `elekLogT.ERROR` with `elekLogT.CONSOLE`.

balandi1 replied:
Done

pradykaushik commented on def/taskUtils.go:
```diff
@@ -105,7 +108,8 @@ func clusterSizeAvgMMMPU(tasks []Task, taskObservation func(task Task) []float64
 				// there is an error in the task config.
-				log.Println(err)
+				elektronLogging.ElektronLog.Log(elekLogT.ERROR, log.ErrorLevel,
+					log.Fields{}, fmt.Sprintf("%s", err))
```
`err.Error()` returns the string representation.

balandi1 replied:
Okay. Will do the change

pradykaushik commented on elektronLogging/ElektronFormatter.go:
```diff
@@ -0,0 +1,51 @@
+	}
+
+	levelColor := f.getColor(entry)
+	level := levelColor.Sprintf("[%s]:", strings.ToUpper(entry.Level.String()))
```
We do not need to color the levels.

pradykaushik commented on elektronLogging/clsfnTaskDistOverheadLogger.go:
```diff
@@ -0,0 +1,50 @@
+	"strings"
+)
+
+type ClsfnTaskDistOverheadLogger struct {
```
Maybe refactor to `ClsfnTaskDistrOverheadLogger` so that it is clear that we are intending "Distribution" and now "Distance"?

balandi1 replied:
Done

pradykaushik commented on elektronLogging/clsfnTaskDistOverheadLogger.go:
```diff
@@ -0,0 +1,50 @@
+
+	tskDistLogPrefix := strings.Join([]string{prefix, config.TaskDistConfig.FilenameExtension}, "")
+	if logDir != "" {
+		tskDistLogPrefix = strings.Join([]string{logDir, tskDistLogPrefix}, "/")
```
You should use [filepath.Join(...)](https://golang.org/src/path/filepath/path.go?s=5893:5925#L199) to keep it generic.

balandi1 replied:
Okay. Sure

pradykaushik commented on elektronLogging/clsfnTaskDistOverheadLogger.go:
```diff
@@ -0,0 +1,50 @@
+	}
+}
+
+func (cLog *ClsfnTaskDistOverheadLogger) SetLogFile(prefix string) {
```
You can possible rewrite this function to something like the one shown below.
```go
func (cLog *ClsfnTaskDistOverheadLogger) SetLogFile(prefix string) {
    filename := prefix + config.TaskDistConfig.FilenameExtension
    if logDir != "" {
        if logFile, err := os.Create(filepath.Join(logDir, filename)); err != nil {
            log.Fatal("Unable to create logFile: ", err)
        } else {
            cLog.LogFileName = logFile
            cLog.AllowOnConsole = config.TaskDistConfig.AllowOnConsole
        }
    }
}
```

balandi1 replied:
Done

pradykaushik commented on elektronLogging/createLogDir.go:
```diff
@@ -0,0 +1,34 @@
+
+func createLogDir(prefix string, startTime time.Time) string {
+
+	// Creating directory to store all logs for this run
```
Extend the comment to show how the format of the log directory name is going to look.

balandi1 replied:
Done

pradykaushik commented on elektronLogging/consoleLogger.go:
```diff
@@ -0,0 +1,47 @@
+		log.Fatal("Unable to create logFile: ", err)
+	} else {
+		cLog.LogFileName = logFile
+		cLog.AllowOnConsole = true
```
I see why you are directly assigning `true`. However, to allow disabling console logs to stdout retrofit this to `cLog.AllowOnConsole = config.ConsoleConfig.AllowOnConsole`.

balandi1 replied:
Sure

pradykaushik commented on elektronLogging/createLogDir.go:
```diff
@@ -0,0 +1,34 @@
+package elektronLogging
+
+import (
+	logrus "github.com/sirupsen/logrus"
```
Keep the alias for logrus import consistent. So, refactor this to "log".

balandi1 replied:
Done

pradykaushik commented on elektronLogging/logger.go:
```diff
@@ -0,0 +1,49 @@
+var config LoggerConfig
+var logger *log.Logger
+var formatter ElektronFormatter
+var ElektronLog *LoggerImpl
```
Logger and Log are synonyms and therefore should not be used interchangeably.

balandi1 replied:
Changed `ElektronLog` to `ElektronLogger`

pradykaushik commented on elektronLogging/logger.go:
```diff
@@ -0,0 +1,49 @@
+
+func BuildLogger(prefix string) {
+
+	// read configuration from yaml
```
Comments should start with capital letters and should end with a full stop.

balandi1 replied:
Done

pradykaushik commented on elektronLogging/logger.go:
```diff
@@ -0,0 +1,49 @@
+	startTime := time.Now()
+	formatter.TimestampFormat = "2006-01-02 15:04:05"
+	formattedStartTime := startTime.Format("20060102150405")
+	GetLogDir(startTime, "_")
```
The second argument to GetLogDir() is a prefix. How is "_" a prefix? This is probably why the log directory name is not ending up prefixed with the one specified with the `-logPrefix` flag.

balandi1 replied:
Done

pradykaushik commented on elektronLogging/consoleLogger.go:
```diff
@@ -0,0 +1,47 @@
+	cLog.SetLogFile(prefix)
+	return cLog
+}
+func (cLog *ConsoleLogger) Log(logType int, level log.Level, logData log.Fields, message string) {
```
Why the pointer receiver? Technically, this function should not lead to change of state.

balandi1 replied:
Yeah, thats right. I will change it

pradykaushik commented on elektronLogging/logger.go:
```diff
@@ -0,0 +1,49 @@
+	logger.SetFormatter(&formatter)
+
+	// create a chain of loggers
+	head := &LoggerImpl{}
```
What is the rationale behind having a dummy node as the head of the chain?

balandi1 replied:
Nothing specific. I had kept `ElektronLog` as `LoggerImpl` type. But I have changed to `ConsoleLogger` rather. Works fine now

pradykaushik commented on elektronLogging/loggerChain.go:
```diff
@@ -0,0 +1,28 @@
+
+type Logger interface {
+	SetNext(logType Logger)
+	Log(logType int, level log.Level, logData log.Fields, message string)
```
As fields are optional, it is better to either have them as the last argument or not have them in the argument list at all.
I would in fact also maintain an internal data structure to store Fields that is cleared only after the log operation passes through the entire chain. You can then add a `WithFields(...)` method that wraps around `logrus.WithFields(...)`.
I have created [this sample code](https://play.golang.org/p/RILTxcZtT2w) for reference. The sample code also shows how formatted strings are corresponding args can be passed and we can wrap around `logrus.Logf(...)` for that.

balandi1 replied:
Done

pradykaushik commented on elektronLogging/loggerConfig.go:
```diff
@@ -0,0 +1,68 @@
+	SchedTraceConfig struct {
+		Enabled             bool   `yaml:"enabled"`
+		FilenameExtension   string `yaml:"filenameExtension"`
+		EnableColumnHeaders bool   `yaml:"enableColumnHeaders"`
```
Are column headers being added if enabled?

balandi1 replied:
That was for future purpose. But I have removed it now

pradykaushik commented on elektronLogging/loggerConfig.go:
```diff
@@ -0,0 +1,68 @@
+		AllowOnConsole      bool   `yaml:"allowOnConsole"`
+	} `yaml:"sps"`
+
+	TaskDistConfig struct {
```
Refactor to TaskDistrConfig for semantics.

balandi1 replied:
Done

pradykaushik commented on elektronLogging/loggerConfig.go:
```diff
@@ -0,0 +1,68 @@
+		FilenameExtension   string `yaml:"filenameExtension"`
+		EnableColumnHeaders bool   `yaml:"enableColumnHeaders"`
+		AllowOnConsole      bool   `yaml:"allowOnConsole"`
+	} `yaml:"clsfnTaskDistOverhead"`
```
ditto.

balandi1 replied:
Done

pradykaushik commented on elektronLogging/loggerConfig.go:
```diff
@@ -0,0 +1,68 @@
+
+	yamlFile, err := ioutil.ReadFile(elekEnv.LogConfigYaml)
+	if err != nil {
+		log.Printf("yamlFile.Get err   #%v ", err)
```
The error message does not do justice in indicating the reason for the error.

balandi1 replied:
Changed the error message

pradykaushik commented on elektronLogging/loggerConfig.go:
```diff
@@ -0,0 +1,68 @@
+	}
+	err = yaml.Unmarshal(yamlFile, c)
+	if err != nil {
+		log.Fatalf("Unmarshal: %v", err)
```
ditto.

balandi1 replied:
Done

pradykaushik commented on elektronLogging/loggerChain.go:
```diff
@@ -0,0 +1,28 @@
+type LoggerImpl struct {
+	Type           int
+	AllowOnConsole bool
+	LogFileName    *os.File
```
Refactor to `LogFile` as it is a pointer to os.File and not string.

balandi1 replied:
Done

pradykaushik commented on elektronLogging/pcpLogger.go:
```diff
@@ -0,0 +1,50 @@
+	"strings"
+)
+
+type PcpLogger struct {
```
Refactor to `PCPLogger`.

balandi1 replied:
Done

pradykaushik commented on elektronLogging/pcpLogger.go:
```diff
@@ -0,0 +1,50 @@
+	pLog := &PcpLogger{}
+	pLog.Type = logType
+	pLog.SetLogFile(prefix)
+	return pLog
```
Why not just do this?
```go
return &PCPLogger{ 
    LoggerIml{
        Type: logType
        LogFile: CreateLogFile(prefix),
    }
}
```

Where `func CreateLogFile(prefix string) string` replaces `func SetLogFile(prefix string)`.

pradykaushik commented on environment/env.go:
```diff
@@ -26,3 +26,5 @@ var RaplPassword = "RAPL_PSSWD"
 // Location of the script that sets the powercap value for a host.
 var RaplThrottleScriptLocation = "RAPL_PKG_THROTTLE_SCRIPT_LOCATION"
+
+var LogConfigYaml = "logConfig.yaml"
```
This file is intended to only store environment variables.

balandi1 replied:
Removed it from here. Added a flag in scheduler.go

pradykaushik commented on logConfig.yaml:
```diff
@@ -0,0 +1,38 @@
+  minLogLevel: debug
+  enableColumnHeaders: true
+pcp:
+    enabled: true
```
Inconsistent spacing.

balandi1 replied:
Done

pradykaushik commented on logConfig.yaml:
```diff
@@ -0,0 +1,38 @@
+    allowOnConsole: true
+    
+
+# format.
```
unnecessary comment.

balandi1 replied:
Removed

pradykaushik commented on schedulers/helpers.go:
```diff
@@ -129,13 +134,13 @@ func WithPCPLog(pcpLog chan struct{}) SchedulerOptions {
 }
 
-func WithLoggingChannels(lmt chan elekLogDef.LogMessageType, msg chan string) SchedulerOptions {
+/*func WithLoggingChannels(lmt chan elekLogDef.LogMessageType, msg chan string) SchedulerOptions {
```
Remove commented code if not required.

balandi1 replied:
Done

balandi1 committed:
Changed alias elekLogT to elekLogTypes

balandi1 committed:
Encapsulated log directory name into a struct

balandi1 committed:
Renamed alias log to elekLog

balandi1 committed:
Revert previous commit. Renamed wrong alias

balandi1 committed:
Changed alias for elektronLogging to elekLog

balandi1 committed:
Removed ERROR, GENERAL, SUCCESS, WARNING log types

balandi1 committed:
Few code improvements and refactoring

pradykaushik commented on elektronLogging/ElektronFormatter.go:
```diff
@@ -0,0 +1,51 @@
+	var formattedFields []string
+	for key, value := range entry.Data {
+		formattedFields = append(formattedFields,
+			strings.Join([]string{key, value.(string)}, "="))
```
Line 43 and 44 can fit in one line. We currently do not have any style linting setup.

balandi1 replied:
Okay

pradykaushik commented on schedulers/base.go:
```diff
@@ -252,179 +248,169 @@ func (s *BaseScheduler) StatusUpdate(driver sched.SchedulerDriver, status *mesos
-	lmt := elekLogDef.GENERAL
-	msgColor := elekLogDef.LogMessageColors[lmt]
-	var msg string
+	lmt := elekLogTypes.CONSOLE
```
refactor variable name to `logMsgType`. In fact, you don't need to have this variable at all. Instead, you can just use `elekLogTypes.CONSOLE` directly.
Make the changes in subsequent occurrences too.

balandi1 replied:
Okay. Done

pradykaushik commented on schedulers/base.go:
```diff
@@ -252,179 +248,169 @@ func (s *BaseScheduler) StatusUpdate(driver sched.SchedulerDriver, status *mesos
-	msg := msgColor.Sprintf("Watts considered for task[%s] and host[%s] = %f Watts",
-		ts.Name, host, wattsToConsider)
-	s.Log(lmt, msg)
+	lmt := elekLogTypes.CONSOLE
```
ditto

pradykaushik commented on schedulers/base.go:
```diff
@@ -252,179 +248,169 @@ func (s *BaseScheduler) StatusUpdate(driver sched.SchedulerDriver, status *mesos
-	msgColor := elekLogDef.LogMessageColors[lmt]
-	msg := msgColor.Sprintf("Received %d resource offers", len(offers))
-	s.Log(lmt, msg)
+	lmt := elekLogTypes.CONSOLE
```
ditto

pradykaushik commented on schedulers/base.go:
```diff
@@ -252,179 +248,169 @@ func (s *BaseScheduler) StatusUpdate(driver sched.SchedulerDriver, status *mesos
-	msg := msgColor.Sprintf("DECLINING OFFER for host[%s]... "+
-		"No tasks left to schedule", offer.GetHostname())
-	s.Log(lmt, msg)
+	lmt := elekLogTypes.CONSOLE
```
ditto

pradykaushik commented on schedulers/base.go:
```diff
@@ -252,179 +248,169 @@ func (s *BaseScheduler) StatusUpdate(driver sched.SchedulerDriver, status *mesos
-	msgColor := elekLogDef.LogMessageColors[lmt]
-	msg := msgColor.Sprintf("Number of tasks still Running = %d", s.tasksRunning)
-	s.Log(lmt, msg)
+	lmt := elekLogTypes.CONSOLE
```
ditto

pradykaushik commented on schedulers/base.go:
```diff
@@ -252,179 +248,169 @@ func (s *BaseScheduler) StatusUpdate(driver sched.SchedulerDriver, status *mesos
 func (s *BaseScheduler) LogCoLocatedTasks(slaveID string) {
-	lmt := elekLogDef.GENERAL
-	msgColor := elekLogDef.LogMessageColors[lmt]
+	lmt := elekLogTypes.CONSOLE
```
ditto

pradykaushik commented on schedulers/base.go:
```diff
@@ -252,179 +248,169 @@ func (s *BaseScheduler) StatusUpdate(driver sched.SchedulerDriver, status *mesos
-	msgColor := elekLogDef.LogMessageColors[lmt]
-	msg := msgColor.Sprint("Done scheduling all tasks!")
-	s.Log(lmt, msg)
+	lmt := elekLogTypes.CONSOLE
```
ditto

pradykaushik commented on schedulers/base.go:
```diff
@@ -252,179 +248,169 @@ func (s *BaseScheduler) StatusUpdate(driver sched.SchedulerDriver, status *mesos
 	offerResources ...interface{}) {
-	lmt := elekLogDef.WARNING
-	msgColor := elekLogDef.LogMessageColors[lmt]
+	lmt := elekLogTypes.CONSOLE
```
ditto

pradykaushik commented on schedulers/base.go:
```diff
@@ -252,179 +248,169 @@ func (s *BaseScheduler) StatusUpdate(driver sched.SchedulerDriver, status *mesos
-	msgColor := elekLogDef.LogMessageColors[lmt]
-	msg := msgColor.Sprintf("OFFER RESCINDED: OfferID = %s", offerID)
-	s.Log(lmt, msg)
+	lmt := elekLogTypes.CONSOLE
```
ditto

pradykaushik commented on schedulers/base.go:
```diff
@@ -252,179 +248,169 @@ func (s *BaseScheduler) StatusUpdate(driver sched.SchedulerDriver, status *mesos
-	msgColor := elekLogDef.LogMessageColors[lmt]
-	msg := msgColor.Sprintf("SLAVE LOST: SlaveID = %s", slaveID)
-	s.Log(lmt, msg)
+	lmt := elekLogTypes.CONSOLE
```
ditto

pradykaushik commented on schedulers/base.go:
```diff
@@ -252,179 +248,169 @@ func (s *BaseScheduler) StatusUpdate(driver sched.SchedulerDriver, status *mesos
-	msgColor := elekLogDef.LogMessageColors[lmt]
-	msg := msgColor.Sprintf("EXECUTOR LOST: ExecutorID = %s, SlaveID = %s", executorID, slaveID)
-	s.Log(lmt, msg)
+	lmt := elekLogTypes.CONSOLE
```
ditto

pradykaushik commented on schedulers/base.go:
```diff
@@ -252,179 +248,169 @@ func (s *BaseScheduler) StatusUpdate(driver sched.SchedulerDriver, status *mesos
-	msgColor := elekLogDef.LogMessageColors[lmt]
-	msg := msgColor.Sprintf("Received Framework message from executor [%s]: %s", executorID, message)
-	s.Log(lmt, msg)
+	lmt := elekLogTypes.CONSOLE
```
ditto

pradykaushik commented on schedulers/base.go:
```diff
@@ -252,179 +248,169 @@ func (s *BaseScheduler) StatusUpdate(driver sched.SchedulerDriver, status *mesos
-	msgColor := elekLogDef.LogMessageColors[lmt]
-	msg := msgColor.Sprintf("MESOS ERROR: %s", err)
-	s.Log(lmt, msg)
+	lmt := elekLogTypes.CONSOLE
```
ditto

pradykaushik commented on schedulers/base.go:
```diff
@@ -252,179 +248,169 @@ func (s *BaseScheduler) StatusUpdate(driver sched.SchedulerDriver, status *mesos
+	lmt := elekLogTypes.CONSOLE
+	elekLog.ElektronLog.Log(lmt,
+		log.InfoLevel,
+		log.Fields{"master": fmt.Sprintf("%s", masterInfo)}, "Framework re-registered")
```
type mismatch. You might want to use either %v or you can use [masterInfo.GetHostname()](https://github.com/mesos/mesos-go/blob/29de6ff97b48c29cb5ac07029ed75186e5ba0eed/api/v0/mesosproto/mesos.pb.go#L1581).

balandi1 replied:
Done

pradykaushik commented on schedulers/base.go:
```diff
@@ -252,179 +248,169 @@ func (s *BaseScheduler) StatusUpdate(driver sched.SchedulerDriver, status *mesos
+	lmt := elekLogTypes.CONSOLE
+	elekLog.ElektronLog.Log(lmt,
+		log.WarnLevel,
+		log.Fields{}, "Framework disconnected with master")
```
Will fit on a single line.

balandi1 replied:
Done

pradykaushik commented on schedulers/helpers.go:
```diff
@@ -23,21 +23,26 @@ import (
 
 func coLocated(tasks map[string]bool, s BaseScheduler) {
 
 	for task := range tasks {
```
Here, `task` would be the index. 
```go
    for _, task := range tasks {
        // rest of the code.
    }
```

balandi1 replied:
Okay

pradykaushik commented on utilities/offerUtils/offerUtils.go:
```diff
@@ -88,12 +90,14 @@ func UpdateEnvironment(offer *mesos.Offer) {
 	if _, ok := constants.Hosts[host]; !ok {
-		log.Printf("New host detected. Adding host [%s]", host)
+		elekLog.ElektronLog.Log(elekLogTypes.CONSOLE, log.InfoLevel,
+			log.Fields{"Adding host": fmt.Sprintf("%s", host)}, "New host detected")
```
As `host` is a string, you don't need to use `fmt.Sprintf`.

balandi1 replied:
Done

pradykaushik commented on utilities/offerUtils/offerUtils.go:
```diff
@@ -88,12 +90,14 @@ func UpdateEnvironment(offer *mesos.Offer) {
 		class := PowerClass(offer)
-		log.Printf("Registering the power class... Host [%s] --> PowerClass [%s]", host, class)
+		elekLog.ElektronLog.Log(elekLogTypes.CONSOLE, log.InfoLevel,
+			log.Fields{"host": fmt.Sprintf("%s", host), "PowerClass": fmt.Sprintf("%s", class)}, "Registering the power class...")
```
ditto for `host` and `class`.

balandi1 replied:
Done

balandi1 committed:
Code Refactoring and modifications

ridv commented:
Thanks @balandi1 .This was a big undertaking and I really appreciate the effort you're putting towards getting it merged. Looks good to me as soon as all of @pradykaushik 's concerns are addressed. Very excited to have you be part of the list of contributors once this is merged.

balandi1 committed:
Refactoring Code

Renamed PcpLogger, SetLogFile(), LogFileName. Changed error messages in loggerConfig.

balandi1 committed:
Flag for Logger Config yaml file name

Removed yaml filename from env.go and added its flag to scheduler.go

balandi1 committed:
Minor code improvements

Removed commented code. Resolved spacing issue in yaml.

balandi1 committed:
Code refactoring

Removed lmt usage in base.go.
Removed unneccessary fmt.Sprintf() usage.

balandi1 committed:
Removed dummy head logger

Replaced dummy head logger instance with ConsoleLogger
instance in the chain.

balandi1 committed:
Code Refactoring

Modified comments. Organized imports.

balandi1 committed:
Handled ennabling/disabling log types

Added enabled check while creating log files and logging.
Do not allow to create log file or log if the log type is disabled.

balandi1 committed:
Removed Log Color Codes

Since the color codes were displayed in an inappropriate manner,
removed them from all logs
.

balandi1 committed:
Added functions to logging library

Added Logf() and WithFields() functions.
Logf() for logging formatted messages.
WithFields() for handling optional log fields.

balandi1 committed:
Minor bug fixes

balandi1 committed:
Minor code style fixes

Removed the global logger instance and kept it in LoggerImpl
struct.
Changed access level for members of LoggerImpl.

balandi1 committed:
Refactored Code

Moved the 'logDir' and 'logger' global variables into 'loggerImpl' struct
Renamed struct names and interface name

pradykaushik committed:
refactored elektronLogging to logging

pradykaushik committed:
refactored Dist to Distr in clsfnTaskDistOverheadLogger.go

pradykaushik committed:
move cfg to loggers + refactor + log fn wrappers

1. Instead of maintaining a global config, each specialized logger
now stores its config.
2. Refactored logInterface to elektronLogger.
3. Refactored loggerImpl to baseElektronLogger to be consistent with
the rest of the code base.
4. Wrapped elektronLogger#Log(...) and elektronLogf(...) so that we
do not have to use the instance of elektronLogger everytime we want
to log. Instead, we just do logging.Log(...) or logging.Logf(...).
5. Wrapped elektronLogger#WithFields(...) and
elektronLogger#WithField(...).
6. Refactored codebase to adhere to the changes.

pradykaushik committed:
refactored logging/types imports to dot imports

pradykaushik committed:
Used WithField() and Logf() where required.

Used elektronLogger#WithField(...) for single key-value pair.
Used elektronLogger#Logf(...) where formatted string is required.

pradykaushik committed:
passing strings instead of pointers as values.

pradykaushik committed:
fixed visibility of structs.

balandi1 committed:
Merge pull request #1 from pradykaushik/switch-to-logrus-contrib

move config to loggers, refactoring, wrapping Log() and Logf()

balandi1 committed:
Added functions to access members

Functions to access all members of baseElektronLogger.config

pradykaushik commented on logConfig.yaml:
```diff
@@ -0,0 +1,32 @@
+  filenameExtension: .pcplog
+  allowOnConsole: false
+schedWindow:
+  enabled: true
```
disable schedWindow logging in the log config file that is kept within the repository.

balandi1 replied:
Done

pradykaushik commented on logConfig.yaml:
```diff
@@ -0,0 +1,32 @@
+  filenameExtension: _schedWindow.log
+  allowOnConsole: true
+clsfnTaskDistrOverhead:
+  enabled: true
```
ditto

balandi1 replied:
Done

pradykaushik commented on logging/elektronFormatter.go:
```diff
@@ -0,0 +1,36 @@
+	TimestampFormat string
+}
+
+func (f elektronFormatter) Format(entry *log.Entry) ([]byte, error) {
```
Does the formatter adhere to the format specified in the config provided?

balandi1 replied:
No, actually as per our discussion earlier you had suggested to implement a decorator pattern for using the format specified in config. So currently, it  has a fixed format and there is technically no use of the format specified in config file.

pradykaushik commented on logging/logger.go:
```diff
@@ -0,0 +1,150 @@
+
+type elektronLogger interface {
+	setNext(next elektronLogger)
+	isEnabled() bool
```
`isEnabled()` does not have to be part of the interface.

balandi1 replied:
Yes. Removed it from there

pradykaushik commented on logging/logger.go:
```diff
@@ -0,0 +1,150 @@
+type elektronLogger interface {
+	setNext(next elektronLogger)
+	isEnabled() bool
+	isAllowedOnConsole() bool
```
ditto

balandi1 replied:
Done

pradykaushik commented on logging/logger.go:
```diff
@@ -0,0 +1,150 @@
+	setNext(next elektronLogger)
+	isEnabled() bool
+	isAllowedOnConsole() bool
+	getFilenameExtension() string
```
ditto

balandi1 replied:
Done

pradykaushik commented on logging/loggerConfig.go:
```diff
@@ -0,0 +1,65 @@
+	Format []string `yaml:"format"`
+}
+
+func GetConfig(logConfigFilename string) (*loggerConfig, error) {
```
does this need to be exported outside the package?

balandi1 replied:
No, not needed. Changed it to `getConfig` .

pradykaushik commented on scheduler.go:
```diff
@@ -76,6 +78,7 @@ func init() {
 	flag.BoolVar(fixSchedWindow, "fixSw", false, "Fix the size of the scheduling window that every deployed scheduling policy should schedule, provided switching is enabled (shorthand).")
 	flag.IntVar(schedWindowSize, "swSize", 200, "Size of the scheduling window if fixSchedWindow is set (shorthand).")
 	flag.StringVar(schedPolSwitchCriteria, "spsCriteria", "taskDist", "Scheduling policy switching criteria (shorthand).")
+	flag.StringVar(logConfigFilename, "lgConfigName", "logConfig.yaml", "Log Configuration file name (shorthand).")
```
change the shorthand to `logCfg`

balandi1 replied:
Done

pradykaushik commented on scheduler.go:
```diff
@@ -228,28 +222,25 @@ func main() {
-	go logger.Listen(logMType, logMsg)
+
+	// Build Logger.
+	if err := elekLog.BuildLogger(*pcplogPrefix, *logConfigFilename); err != nil {
```
In the future, we can have the log prefix to be part of the log config.

balandi1 replied:
Oh, okay.

pradykaushik commented on utilities/schedUtils/schedUtils.go:
```diff
@@ -77,8 +78,9 @@ func (s *fillNextOfferCycle) apply(taskQueue []def.Task) (int, int) {
 		for i := *task.Instances; i > 0; i-- {
-			log.Printf("Checking if Instance #%d of Task[%s] can be scheduled "+
-				"during the next offer cycle...", i, task.Name)
+			elekLog.Logf(CONSOLE, log.InfoLevel,
```
can fit in two lines.

balandi1 replied:
Done

pradykaushik commented:
Overall, LGTM.
I have mentioned a few nit picks plus a formatting related fix.
Once those are fixed, I think this is ready to be merged in.
Thank you for undertaking a non-trivial task. This was an important addition to elektron.

balandi1 committed:
Code and formatting fixes

Changed the shorthand for log config flag.
Fixed formatting as mentioned.
Removed unnecessary function declarations from elektronLogger interface.
Disabled schedWindow and clsfnTaskDistrOverhead in config file.

balandi1 committed:
Function renaming

Renamed 'GetConfig()' to 'getConfig()', so that it is not exposed
outside the package.

balandi1 committed:
Updated log format for schedTrace

balandi1 committed:
Updated log format for SPS

balandi1 committed:
Updated log format for SCHED_WINDOW

balandi1 committed:
Minor format fix

balandi1 committed:
Updated log format for CLSFN_TASKDISTR_OVERHEAD

balandi1 committed:
Updated log format for PCP

balandi1 committed:
Renamed log type

Renamed CLSFN_TASKDIST_OVERHEAD to CLSFN_TASKDISTR_OVERHEAD

balandi1 committed:
Added documentation for log configurations

balandi1 committed:
Removed format from logConfig.yaml

Not used anywhere

balandi1 committed:
Merge branch 'master' of https://github.com/balandi1/elektron

balandi1 committed:
Added license to the new go files
